### PR TITLE
Qute - support Mutiny Uni directly in the templates

### PR DESCRIPTION
**Pull request by mkouba**
- also mention the special handling of CompletionStage and Uni in docs

### CODE REVIEW

File: extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/QuteProcessor.java [hunk 1]
```diff
@@ -1193,14 +1186,59 @@ static Type extractMatchType(Set<Type> closure, DotName matchName, Function<Type
...
+        void autoExtractType() {
+            boolean hasCompletionStage = ValueResolverGenerator.hasCompletionStageInTypeClosure(clazz, index);
+            boolean hasUni = hasCompletionStage ? false
+                    : ValueResolverGenerator.hasClassInTypeClosure(clazz, Names.UNI, index);
+            if (hasCompletionStage || hasUni) {
+                Set<Type> closure = Types.getTypeClosure(clazz, Types.buildResolvedMap(
+                        getParameterizedTypeArguments(), getTypeParameters(), new HashMap<>(), index), index);
+                Function<Type, Type> firstParamType = t -> t.asParameterizedType().arguments().get(0);
+                // CompletionStage<List<Item>> => List<Item>
+                // Uni<List<String> => List<String>
```
Comment (cescoffier) [inline extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/QuteProcessor.java:1237]: Missing a > in the comment

**Review by cescoffier — APPROVED** [review]
Nice!

File: docs/src/main/asciidoc/qute-reference.adoc [hunk 2]
```diff
@@ -440,6 +440,13 @@ Virtual method parameters can be "nested" virtual method invocations.
 ----
 <1> `item.calculateDiscount(10)` is evaluated first and then passed as an argument to `item.subtractPrice()`.
 
+==== Evaluation of `CompletionStage` and `Uni` Objects
+
+Objects that implement `java.util.concurrent.CompletionStage` and `io.smallrye.mutiny.Uni` are evaluated in a special way.
+If a part of an expressions resolves to a `CompletionStage`, the resolution continues once this stage is completed and the next part of the expression (if any) is evaluated against the result of the completed stage.
+For example, if there is an expression `{foo.size}` and `foo` resolves to `CompletionStage<List<String>>` then `size` is resolved against the completed result, i.e. `List<String>`.
+If a part of an expressions resolves to a `Uni`, a `CompletionStage` is first created from `Uni` using `Uni#subscribeAsCompletionStage()` and then evaluated as described above.
```
Comment (gastaldi) [inline docs/src/main/asciidoc/qute-reference.adoc:448]:
```suggestion
If a part of an expression resolves to a `Uni`, a `CompletionStage` is first created from `Uni` using `Uni#subscribeAsCompletionStage()` and then evaluated as described above.
```

File: docs/src/main/asciidoc/qute-reference.adoc [hunk 3]
```diff
@@ -440,6 +440,13 @@ Virtual method parameters can be "nested" virtual method invocations.
 ----
 <1> `item.calculateDiscount(10)` is evaluated first and then passed as an argument to `item.subtractPrice()`.
 
+==== Evaluation of `CompletionStage` and `Uni` Objects
+
+Objects that implement `java.util.concurrent.CompletionStage` and `io.smallrye.mutiny.Uni` are evaluated in a special way.
+If a part of an expressions resolves to a `CompletionStage`, the resolution continues once this stage is completed and the next part of the expression (if any) is evaluated against the result of the completed stage.
```
Comment (gastaldi) [inline docs/src/main/asciidoc/qute-reference.adoc:446]:
```suggestion
If a part of an expression resolves to a `CompletionStage`, the resolution continues once this stage is completed and the next part of the expression (if any) is evaluated against the result of the completed stage.
```

**Pull request merged**